File: Queue/Circular Queue/Array-Based/CircQueueArray.java
public class CircQueueArray{

    private int elements[];
    private int FRONT;
    private int REAR;

    //Initialization
    CircQueueArray(int size){

        elements = new int[size];
        FRONT = -1;
        REAR = -1;

    }

    //Check whether the Queue is Full or not
    public boolean isFull(){
        return FRONT == REAR + 1 || FRONT == 0 && REAR == elements.length - 1;
    }

    //Check whether the Queue is Empty or not
    public boolean isEmpty(){
        return FRONT == -1 && REAR == -1;
    }

    public void display(){

        if(isEmpty()){
            System.out.println("Queue is Empty!");
            return;
        }

        for(int i = 0; i < elements.length; i++)
            if(i != elements.length - 1)
                System.out.print(elements[i] + " <- ");
            else
                System.out.print(elements[i]);

        System.out.println();
    }


    public void Enqueue(int x){

        System.out.println("\nBefore Enqueue: FRONT = " + FRONT + " and REAR = " + REAR);

        if(isFull()){
            System.out.println("Queue is Full!");
            return;
        }

        if(isEmpty()){

            FRONT++; REAR++;

            elements[FRONT] = x;

        } else {

            REAR = (REAR + 1) % elements.length;
            elements[REAR] = x;

        }

        System.out.println("\nAfter Enqueue: FRONT = " + FRONT + " and REAR = " + REAR);

    }

    //Dequeue
    public void Dequeue(){

        System.out.println("\nBefore Enqueue: FRONT = " + FRONT + " and REAR = " + REAR);

        if(isEmpty()){
            System.out.println("Queue is Empty!");
            return;
        }

        if(FRONT == REAR){

            elements[FRONT++] = 0;
            FRONT = -1;
            REAR = -1;

        } else

            elements[FRONT++] = 0;

        System.out.println("\nAfter Enqueue: FRONT = " + FRONT + " and REAR = " + REAR);

    }

    
}

File: Queue/Circular Queue/Array-Based/CircQueueArray.py
class CircQueueArray:

    def __init__(self, size):
        self.MAX = size
        self.elements = [0] * size
        self.FRONT = -1
        self.REAR = -1


    #Check whether the Queue is Empty or not
    def isEmpty(self):
        return self.FRONT == -1 and self.REAR == -1

    #Check whether the Queue is Full or not
    def isFull(self):
        return self.FRONT == self.REAR + 1 or self.FRONT == 0 and self.REAR == self.MAX - 1;

    #Display
    def display(self):
        print(self.elements)


    #Enqueue
    def Enqueue(self, x):

        print("\nBefore Enqueue: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))

        if self.isFull():
            print("Queue is Full!")
            return

        if self.isEmpty():

            self.FRONT += 1
            self.REAR += 1

            self.elements[self.FRONT] = x

        else:

            self.REAR = (self.REAR + 1) % self.MAX
            self.elements[self.REAR] = x
        
        print("\nAfter Enqueue: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))


    #Dequeue
    def Dequeue(self):

        print("\nBefore Dequeue: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))

        if self.isEmpty():
            print("Queue is Empty!")
            return

        if self.FRONT == self.REAR:

            self.FRONT += 1
            self.elements[self.FRONT] = 0
            self.FRONT = -1
            self.REAR = -1

        else:

            self.elements[self.FRONT] = 0
            self.FRONT += 1
        
        print("\nAfter Dequeue: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))
